block/block_service.proto

```diff
@@ -0,0 +1,710 @@
+         * not found.<br/>
+         * The client MAY retry the request; if this result is repeated the
+         * request SHOULD be directed to a different block node server.
+         */
+        READ_BLOCK_NOT_FOUND = 3;
+
+        /**
+         * The requested block is not available on this block node server.<br/>
+         * The client SHOULD send a `serverStatus` request to determine the
+         * lowest and highest block numbers available at this block node server.
+         */
+        READ_BLOCK_NOT_AVAILABLE = 4;
+    }
+}
+
+// ---- ReadBlockStream RPC ----
+
+
+/**
+ * A request to stream block items from block node to a client.
+ *
+ * The block node SHALL respond to this request with a stream of
+ * `ReadStreamResponse` messages.<br/>
+ * The block node SHALL stream the full contents of the blocks requested.<br/>
+ * The block items SHALL be streamed in order originally produced within
+ * a block.<br/>
+ * The blocks SHALL be streamed in ascending order by `block_number`.<br/>
+ * The block node SHALL end the stream when the last requested block has
+ * been sent.<br/>
+ * The block node SHALL end the stream with a response code status of SUCCESS
+ * when the stream is complete.<br/>
+ * The client SHOULD call the `serverStatus` rpc prior to constructing this
+ * request to determine the available start and end blocks.
+ *
+ * > REVIEW NOTE
+ * >> This lacks application-level error detection. If the client receives
+ * >> a bad message (e.g. due to networking errors) it must await the full
+ * >> stream, then determine which item failed, then request at least one,
+ * >> and possibly many, blocks again.
+ * >
+ * >> Should we offer a more callback-style interface where the client
+ * >> offers a gRPC endpoint that implements `publishBlockStream`, as
+ * >> specified below? That would allow the client to perform ack/nack
+ * >> on a per-item basis and we could immediately retransmit only
+ * >> failed items (reducing waste).
+ */
+message SubscribeStreamRequest {
+    /**
+     * A block number
+     * <p>
+     * This SHALL be the block number of the first block returned.<br/>
+     * This field MUST be less than or equal to the latest available
+     * block number.
+     */
+    uint64 start_block_number = 1;
+
+    /**
+     * A block number
+     * <p>
+     * This value SHALL be the block number of the last block returned.<br/>
+     * This field MUST NOT be less than `start_block_number`.<br/>
+     * This SHOULD be a block number that is immediately available from the
+     * block node.<br/>
+     * A block node SHALL continue to stream blocks until the last requested
+     * block is transmitted.<br/>
+     * A block node implementation MAY reject a request for a block that is not
+     * yet available.<br/>
+     * A block node implementation MAY accept future block numbers.
+     * Block node implementations SHOULD charge increased fees for such
+     * "future" streams.
+     */
+    uint64 end_block_number = 2;
```

Comment:
We should consider adding a limit field. A subscriber would define one or the other if at all

Reply:
What would a limit look like, if not the block number at which to end the stream?

Reply:
The number of block.
However since we're streaming block Items the initial limit idea doesn't apply as much